src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,258 @@
+        new_actions = action_dists.rsample()
+        log_pi = action_dists.log_prob(new_actions)
+
+        # tune the coefficient alpha
+        if self.use_automatic_entropy_tuning:
+            alpha_loss = (-(self.log_alpha)* (log_pi.detach() + self.target_entropy)).mean()
+            self.alpha_optimizer.zero_grad()
+            alpha_loss.backward()
+            self.alpha_optimizer.step()
+            tabular.record("alpha", torch.exp(self.log_alpha.detach()).item())
+            alpha = self.log_alpha.detach().exp().item()
+        else:
+            alpha = self.alpha
+        min_q_new_actions = torch.min(self.qf1(torch.Tensor(obs), torch.Tensor(new_actions)), 
+                            self.qf2(torch.Tensor(obs), torch.Tensor(new_actions)))
+        policy_objective = ((alpha * log_pi) - min_q_new_actions.flatten()).mean()
+        tabular.record("policy_loss", policy_objective.item())
+
+        '''
+        QF Loss
+        '''
+        q1_pred = self.qf1(torch.Tensor(obs), torch.Tensor(actions))
+        q2_pred = self.qf2(torch.Tensor(obs), torch.Tensor(actions))
+        new_next_actions_dist = self.policy(torch.Tensor(next_obs))
+        new_next_actions = new_next_actions_dist.rsample()
+        new_log_pi = new_next_actions_dist.log_prob(new_next_actions)
+
+        target_q_values = torch.min(
+            self.target_qf1(torch.Tensor(next_obs), new_next_actions),
+            self.target_qf2(torch.Tensor(next_obs), new_next_actions)
+        ).flatten() - alpha * new_log_pi
+
+        q_target = torch.Tensor(rewards) + (1. - torch.Tensor(terminals)) * self.discount * target_q_values
+        qf1_loss = F.mse_loss(q1_pred.flatten(), q_target.detach())
+        qf2_loss = F.mse_loss(q2_pred.flatten(), q_target.detach())
+
+        self.qf1_optimizer.zero_grad()
+        qf1_loss.backward()
+        self.qf1_optimizer.step()
+
+        self.qf2_optimizer.zero_grad()
+        qf2_loss.backward()
+        self.qf2_optimizer.step()
+
+        self.policy_optimizer.zero_grad()
+        policy_objective.backward()
+        self.policy_optimizer.step()
+
+    def temperature_objective(self, log_pi):
+        """
+        implemented inside optimize_policy
+        """
+        alpha_loss = 0
+        if self.use_automatic_entropy_tuning:
+            alpha_loss = (-(self.log_alpha)* (log_pi.detach() + self.target_entropy)).mean()
+        return alpha_loss
+
+    def actor_objective(self, obs, log_pi, new_actions):
+        alpha = self.log_alpha.detach().exp()
+        min_q_new_actions = torch.min(self.qf1(torch.Tensor(obs), torch.Tensor(new_actions)), 
+                            self.qf2(torch.Tensor(obs), torch.Tensor(new_actions)))
+        policy_objective = ((alpha * log_pi) - min_q_new_actions.flatten()).mean()
+        return policy_objective
+
+    def critic_objective(self,samples):
+        '''
+        QF Loss
+        '''
+        obs = samples["observation"]
```

Comment:
didn't you already do this unpacking somewhere else? why not do it once and then call these functions using the unpacked values?